hal/src/common/sercom/v2/spi_future.rs

```diff
@@ -0,0 +1,469 @@
+    /// interrupts.
+    #[inline]
+    pub fn start(&mut self) {
+        self.ss.assert();
+        self.spi.as_mut().enable_interrupts(Flags::DRE | Flags::RXC);
+    }
+
+    /// Send the next set of bytes from the buffer
+    ///
+    /// This method should be called from the `DRE` interrupt handler. Once all
+    /// bytes of the transaction have been sent, this function will
+    /// automatically disable the `DRE` interrupt.
+    pub fn send(&mut self) -> Result<(), Error> {
+        let _ = self.spi.as_ref().read_flags_errors()?;
+        let buf = self.buf.as_ref();
+        if let Some(buf) = buf.get(self.sent..) {
+            let mut data = buf.into_iter();
+            let mut bytes = [0; 4];
+            let mut iter = bytes.iter_mut();
+            for _ in 0..S::STEP {
+                match (iter.next(), data.next()) {
+                    (Some(b), Some(d)) => *b = *d,
+                    _ => break,
+                }
+            }
+            let word = u32::from_le_bytes(bytes);
+            unsafe { self.spi.as_mut().write_data(word) };
+            self.sent += S::STEP;
+        }
+        if self.sent >= buf.len() {
+            self.spi.as_mut().disable_interrupts(Flags::DRE);
+        }
+        Ok(())
+    }
+
+    /// Received the next set of bytes and write them to the buffer
+    ///
+    /// This method should be called from the `RXC` interrupt handler. Once all
+    /// bytes of the transaction have been received, this function will
+    /// automatically disable the `RXC` interrupt, deassert the SS pin (if
+    /// present), and call the waker (if present).
+    pub fn recv(&mut self) -> Result<(), Error> {
+        let _ = self.spi.as_ref().read_flags_errors()?;
+        let buf = self.buf.as_mut();
+        if self.rcvd < self.sent {
+            let buf = unsafe { buf.get_unchecked_mut(self.rcvd..) };
+            let mut data = buf.into_iter();
+            let word = unsafe { self.spi.as_mut().read_data() };
+            let bytes = word.to_le_bytes();
+            let mut iter = bytes.iter();
+            for _ in 0..S::STEP {
+                match (data.next(), iter.next()) {
+                    (Some(d), Some(b)) => *d = *b,
+                    _ => break,
+                }
+            }
+            self.rcvd += S::STEP;
+        }
+        if self.rcvd >= buf.len() {
+            self.spi.as_mut().disable_interrupts(Flags::RXC);
+            self.ss.deassert();
+            if let Some(waker) = self.waker.take() {
+                waker();
+            }
+        }
+        Ok(())
+    }
+
+    /// Poll the [`SpiFuture`]
+    ///
+    /// This function will return [`Poll::Pending`] until all bytes have been
+    /// received. When the transaction is complete, it will return
+    /// [`Poll::Ready`] with a reference to the completed buffer.
+    #[inline]
+    pub fn poll(&self) -> Poll<&[u8]> {
+        let buf = self.buf.as_ref();
+        if self.rcvd >= buf.len() {
```

Comment:
Shouldn't this read and all writes to `self.rcvd` be volatile?

Reply:
Hmm. I don't think so. But maybe there's something subtle I'm missing.

I think you're suggesting that multiple reads of `self.rcvd` by repeated calls to `poll` would be optimized away, so you would never see updates to `self.rcvd`. Is that right?

If you're using accessing `SpiFuture` from interrupts at different priority levels, then it would need to be wrapped in a `Mutex` of some type. In RTIC, you would `poll` from a closure passed to the `Mutex`. There's no `Mutex` type in `core`, so I'm not sure what you would do there. But any solution would probably look like `std::sync::Mutex`, where locking would create a new `MutexGuard`. In either case, it wouldn't make sense to `poll` more than once, since the `Mutex` ensures your access can't be preempted.

If you're accessing `SpiFuture` from interrupts at the same priority level, then they can't preempt each other to begin with. So once again, it doesn't make sense to `poll` more than once.

Is there a scenario I'm missing? Or something I'm misunderstanding?

Reply:
You obviously thought this through a lot more than I have. I am pretty sure that you would need volatile for similar code in C, which I am more used to. I forget that Rust's ownership/safety model prevents these kinds of mistakes.

Reply:
Just because I've thought about it doesn't mean it's correct =D

I'm relying on the assumption that using only safe code should do what I want/expect to happen. I think that's right in this case, but I'm not 100% sure.

Reply:
I still have to learn to trust the Rust compiler more than I can trust any C compiler. Years of writing C taught me to be very careful and it's hard to unlearn. :upside_down_face: